src/OrchardCore.Modules/OrchardCore.Admin/AdminAreaControllerRouteMapper.cs

```diff
@@ -9,33 +9,45 @@ namespace OrchardCore.Admin
 {
     public class AdminAreaControllerRouteMapper : IAreaControllerRouteMapper
     {
-        private readonly string _defaultAreaPattern;
+        private const string _defaultAreaPattern = "{area}/{controller}/{action}/{id?}";
+        private readonly string _adminUrlPrefix;
 
         public int Order => -1000;
 
         public AdminAreaControllerRouteMapper(IOptions<AdminOptions> adminOptions)
         {
-            _defaultAreaPattern = adminOptions.Value.AdminUrlPrefix + "/{area}/{controller}/{action}/{id?}";
+            _adminUrlPrefix = adminOptions.Value.AdminUrlPrefix;
         }
 
         public bool TryMapAreaControllerRoute(IEndpointRouteBuilder routes, ControllerActionDescriptor descriptor)
         {
-            if (descriptor.ControllerName == "Admin" ||
-                descriptor.ControllerTypeInfo.GetCustomAttribute<AdminAttribute>() != null ||
-                descriptor.MethodInfo.GetCustomAttribute<AdminAttribute>() != null
-                )
+            var controllerAttribute = descriptor.ControllerTypeInfo.GetCustomAttribute<AdminAttribute>();
+            var actionAttribute = descriptor.MethodInfo.GetCustomAttribute<AdminAttribute>();
+
+            if (descriptor.ControllerName != "Admin" && controllerAttribute == null && actionAttribute == null)
+            {
+                return false;
+            }
+
+            var pattern = _defaultAreaPattern;
+
+            if (!string.IsNullOrWhiteSpace(actionAttribute?.Template))
+            {
+                pattern = actionAttribute.Template;
+            }
+            else if (!string.IsNullOrWhiteSpace(controllerAttribute?.Template))
             {
-                routes.MapAreaControllerRoute(
-                    name: descriptor.DisplayName,
-                    areaName: descriptor.RouteValues["area"],
-                    pattern: _defaultAreaPattern.Replace("{action}", descriptor.ActionName),
-                    defaults: new { controller = descriptor.ControllerName, action = descriptor.ActionName }
-                );
-
-                return true;
+                pattern = controllerAttribute.Template;
             }
 
-            return false;
+            routes.MapAreaControllerRoute(
```

Comment:
@sarahelsaig 
Usually if the action is Index, we don't append Index to the route, maybe we can add an in condition to allow us to not have to type /index in the URl for index actions.

Reply:
Because of all the possible optional template parts at the end (like `/{id?}`), I think it would be pretty complicated to make it reliable without the danger of accidentally creating conflicts. Also I think `[Route]` doesn't offer this functionality either, so it may be confusing for users.

I have added an action-level route to every Index where a matching route was defined in the startup. For example [this](https://github.com/OrchardCMS/OrchardCore/pull/15251/files#diff-6d5d8927f63774f4760387afbf2b1f8cd536fd8bb5566321f04aaef9b60f7dc5L54-L59) became [this](https://github.com/OrchardCMS/OrchardCore/pull/15251/files#diff-41124286db00cbe90fba652e472f4fa52b39ad3e8fbcb3ffccd77bda7ad18939R93).

Reply:
Okay. I guess we can add the admin attribute on the index action in addition to the controller. The attribute on the action will have a height order I am guess so it would take presence.

Another thing we can support if you up for it is a name property to identify the name of the route. Sometimes it's is helpful to be able to create url using a route name instead of action/controller. Also, I am not sure what descriptor.DisplayName returns but maybe we can use Name property and fallback on descriptor.DisplayName

Reply:
Here is an example of the default `descriptor.DisplayName`:
```
OrchardCore.Settings.Controllers.AdminController.Index (OrchardCore.Settings)
```
Looks like a cleaned-up version of the action's method header. In my opinion this is easier to understand when debugging than the names used in the `routes.MapAreaControllerRoute()` calls I've removed like AdminMenuList, BackgroundTasks, SearchUsers, etc. Is there any name that should be specifically preserved? I've added `AdminAttribute.RouteName` so it's possible.

Also I've prefixed the default name with "Admin Route for " to avoid confusion with the routes created by `DefaultAreaControllerRouteMapper`. Apparently you can have anything in the route name including spaces.

Reply:
`descriptor.DisplayName` is good now since you added the RouteName option for those he will use it. thank you!